Adding devise authentication

plauring opened the pull request:
-more tests for user and devise functionality

RusselViola committed:
Suuuuuuper WIP end of day push, added feature and unit tests for user-auth and fleshed out form structure, waiting for CRUD update

RusselViola committed:
Merge branch 'master' of https://github.com/plauring/review-site-8 into adding-devise-authentication

- Merged With master branch
- Consolidated factory_girl.rb
- Removed:
    `<p class="notice"><%= notice %></p>`
    `<p class="alert"><%= alert %></p>`
  from application.html.erb

RusselViola committed:
fixed factory girl mixup OOPS

RusselViola committed:
Added sign in/out funcitonality, tests WIP

plauring committed:
Finished user auth tests, configuring user side devise

plauring committed:
made some changes suggested by hound

plauring committed:
Tests not working right now because we cant get warden test helper to work. Made some hound edits as well

plauring committed:
-Reconfigured DBCleaner
-Configured WardenTest for feature testing
-Added 'Role' column to Users Table
-Started Test Refactoring for logged in users vs visiting guests WIP

RusselViola committed:
this is wrong now

RusselViola committed:
removed extraneous folder

RusselViola commented on spec/support/factory_girl.rb:
```diff
@@ -1,10 +1,14 @@
+      sequence(:name) { |n| "PartyBoy200#{n}" }
+      sequence(:email) { |n| "party#{n}eva@launchacademy.com" }
+      profile_picture "https://steamcdn-a.akamaihd.net/steamcommunity/public/images/avatars/5c/5cd3cdae01e76fa0ad460e1573ea030a07bb4dd7_full.jpg"
+      sequence(:password) {|n| "123#{n}4567"}
```
DONE

RusselViola commented on spec/support/factory_girl.rb:
```diff
@@ -1,10 +1,14 @@
+    factory :user do
+      sequence(:name) { |n| "PartyBoy200#{n}" }
+      sequence(:email) { |n| "party#{n}eva@launchacademy.com" }
+      profile_picture "https://steamcdn-a.akamaihd.net/steamcommunity/public/images/avatars/5c/5cd3cdae01e76fa0ad460e1573ea030a07bb4dd7_full.jpg"
```
DONE SO DONE

RusselViola commented on spec/support/factory_girl.rb:
```diff
@@ -1,10 +1,14 @@
-    sequence(:email) { |n| "partyboy200#{n}@example.com"}
-  end
+
+    factory :user do
```
DIGGITY DONE YO

RusselViola commented on spec/support/factory_girl.rb:
```diff
@@ -1,10 +1,14 @@
-    sequence(:username) { |n| "PartyBoy200#{n}" }
-    sequence(:email) { |n| "partyboy200#{n}@example.com"}
-  end
+
```
DONE

RusselViola committed:
removed more extraneous folders

RusselViola commented on spec/support/factory_girl.rb:
```diff
@@ -2,8 +2,10 @@
-    sequence(:email) { |n| "partyboy200#{n}@example.com"}
+    sequence(:name) { |n| "PartyBoy200#{n}" }
+    sequence(:email) { |n| "party#{n}eva@launchacademy.com" }
+    profile_picture 'https://steamcdn-a.akamaihd.net/steamcommunity/public/images/avatars/5c/5cd3cdae01e76fa0ad460e1573ea030a07bb4dd7_full.jp'g
```
SO DONE SEE YA LATER HOUND

RusselViola commented on spec/features/06.add_a_review_spec.rb:
```diff
@@ -0,0 +1,47 @@
+  context 'An authenticated user can add a review to a dev tool' do
+    before {
+    user2 = FactoryGirl.create(:user)
+    login_as(user2, :scope => :user, :run_callbacks => false)
```
DONE

RusselViola committed:
Wrote test suite for Admin privelages
started on code to pass admin tests
WIP

RusselViola committed:
- Finished TDD on User and Admin mechanics
- Added special admin versions to views, as well as privelages
- Added Delete methods to contollers
- Restricted normal users from Admin privelages
- Added a secret Nick Alberts page that only the admin can access and even
some admins don't know the path

RusselViola committed:
obey the hound

RusselViola committed:
release the hounds!

RusselViola committed:
UGHGHHGHGHSODIUGHOADILSGH:KAUSG

RusselViola committed:
gosh darn hound

nwalberts commented on app/controllers/reviews_controller.rb:
```diff
@@ -1,9 +1,16 @@
+    @devtool = Devtool.find(params[:devtool_id])
+    if !user_signed_in?
+      flash[:notice] = 'You must be signed in to post a review'
+      redirect_to devtool_path(@devtool.id)
```
You dont need to use `.id` here, simply passing in the object isfine

RusselViola replied:
fixed

nwalberts commented on app/controllers/users_controller.rb:
```diff
@@ -1,3 +1,42 @@
+      end
+   end
+   redirect_to users_path
+  end
```
You shouldnt need to iterate over EVERY user and try deleting them.....that would be bad

You need to grab an id from params, then use that id to delete ONE user. And you probably dont need line 29

Brady-Tatro replied:
Logic fixed

nwalberts commented on app/views/devtools/index.html.erb:
```diff
@@ -1,5 +1,12 @@
-<% @devtools.each do |devtool|%>
-  <li><%= link_to devtool.title, devtool_path(devtool) %></li>
+<% if (current_user.nil? || user_signed_in?) && (current_user.nil? || !current_user.admin?) %>
```
simplify this logic

RusselViola replied:
removed

nwalberts commented on app/views/devtools/show.html.erb:
```diff
@@ -1,15 +1,29 @@
-<h1>HELP</h1>
+<% if (current_user.nil? || user_signed_in?) && (current_user.nil? || !current_user.admin?) %>
```
simplify this logic

RusselViola replied:
removed

nwalberts commented on app/views/devtools/show.html.erb:
```diff
@@ -1,15 +1,29 @@
 
-<h1><%= @devtool.title %></h1>
+    <div> Dev Tool description <%= @devtool.body %></div>
+    <div> Dev Tool's Github <%= @devtool.github_link %></div>
```
this should be a a href/ link

Brady-Tatro replied:
completed

nwalberts commented on app/views/devtools/show.html.erb:
```diff
@@ -1,15 +1,29 @@
+        <li><%= review.title %><%= link_to "delete #{review.title}", devtool_review_path(@devtool), method: :delete %></li>
+        <li><%= review.body %></li>
+        <li> <%= review.rating %></li>
+      <% end %>
```
This is not dry. You repeat a bunch of code. Simply have a small if block that, if the user is admin, shows a delete button.

nwalberts commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,13 @@
+<img src='http://gianthamster.com/wp-content/uploads/2010/03/2010_03_01_05_tCaplinWalk.jpg'/>
```
i would snip this out

Brady-Tatro replied:
snipped

nwalberts commented on spec/features/02.user_signs_in_and_out_spec.rb:
```diff
@@ -0,0 +1,62 @@
+    fill_in "user_password", with: 'party4life'
+    fill_in 'Password Confirmation', with: 'party4life'
+    click_button 'Sign Up'
+  end
```
Instead of doing this, create a user factory and create the user

RusselViola replied:
done

nwalberts commented on spec/features/04.user_views_a_showpage_spec.rb:
```diff
@@ -1,22 +1,32 @@
+
 end
+
+Warden.test_reset!
```
You shouldnt need line 32

RusselViola replied:
line 32 removed

RusselViola replied:
removed

nwalberts commented on spec/features/05.user_views_all_devTools_spec.rb:
```diff
@@ -0,0 +1,13 @@
+    scenario 'guest sees all the development tools' do
+      devtools.each do |devtool|
+        expect(page).to have_content(devtool.title)
+      end
```
Not a huge fan of iterations in a tests. Simply create 2 dev tools, and then make sure their titles are there outside of a loop

RusselViola replied:
iteration removed

RusselViola replied:
finished

nwalberts commented on spec/features/09.user_clicks_link_to_add_devtool_spec.rb:
```diff
@@ -0,0 +1,27 @@
+feature 'user click on add devtool' do
+
```
address this

nwalberts commented on spec/features/09.user_clicks_link_to_add_devtool_spec.rb:
```diff
@@ -0,0 +1,27 @@
+  end
+  context 'As an unauthenticated user' do
+    scenario 'I can click the link because Im an american, but it will only give me an error' do
+      visit root_path
```
as goofy as you are, change this scenario

Brady-Tatro replied:
creativity crushed

nwalberts commented on spec/features/10.admin_signs_in_spec.rb:
```diff
@@ -0,0 +1,69 @@
+      devtools.each do |devtool|
+        expect(page).to have_content('delete')
+      end
+    end
```
you dont need to iterate over every one, simply make sure one devtool has the delete

nwalberts commented on spec/features/10.admin_signs_in_spec.rb:
```diff
@@ -0,0 +1,69 @@
+    end
+    scenario 'I can visit a Dev Tools page and delete it' do
+      tool = FactoryGirl.create(:devtool)
+      visit devtool_path(tool.id)
```
you dont need `.id`, simply pass in the tool object

nwalberts commented on spec/features/10.admin_signs_in_spec.rb:
```diff
@@ -0,0 +1,69 @@
+      visit users_path
+      users.each do |user|
+        expect(page).to have_content(user.name)
+      end
```
just check for two users, no need to iterate over every one

RusselViola committed:
-Refactoring for pull requests. Mainly Tests and views. No changes to models

RusselViola committed:
more refactors for pull request

RusselViola committed:
more hound

RusselViola committed:
houndy hound hound

RusselViola committed:
that
's right, more hound

RusselViola committed:
one more?

RusselViola committed:
last one I promise

RusselViola committed:
i dont even know anymore

RusselViola commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,49 @@
+require 'rails_helper'
+
+describe User do
+  let!(:user) { User.create(name: 'partybody200',email: '2hard2party@example.com',password: '123qwe') }
```
DONE

RusselViola committed:
more hound

RusselViola committed:
one more for hound
